keepsorted/block.go

```diff
@@ -79,41 +88,125 @@ func (f *Fixer) newBlocks(lines []string, offset int, include func(start, end in
+			// At depth n, n != 0, we increase the length of nestedBlocks to be n.
+			// At depth m=n-1, the length of nestedBlocks will initially be n=m+1 (the assertion from above)
+			// and then we trim that down to be length m when we add the nested blocks
+			// to the current block.
+			if len(nestedBlocks) == depth+1 {
+				block.nestedBlocks = nestedBlocks[depth]
+				nestedBlocks = nestedBlocks[0:depth]
+			}
+			if depth == 0 {
+				// Top-level blocks get returned.
+				// Nested blocks are returned via their top-level block.
+				blocks = append(blocks, block)
+			} else {
+				// Otherwise, the current block appears to be nested. Add it to nestedBlocks.
+				for len(nestedBlocks) < depth {
+					nestedBlocks = append(nestedBlocks, nil)
+				}
+				nestedBlocks[depth-1] = append(nestedBlocks[depth-1], block)
+			}
+			// Invariant: len(nestedBlocks) == depth
 		}
 	}
-	if startIndex >= 0 {
-		incompleteBlocks = append(incompleteBlocks, incompleteBlock{startIndex + offset, startDirective})
+	if len(starts) > 0 {
+		for _, st := range starts {
+			incompleteBlocks = append(incompleteBlocks, incompleteBlock{st.index + offset, startDirective})
+		}
+		// There were some unfinished start directives. They might've caused some
+		// blocks to be incorrectly considered nested.
+		for _, nested := range nestedBlocks {
+			blocks = append(blocks, nested...)
+		}
 	}
 
 	return blocks, incompleteBlocks
 }
 
 func (b block) sorted() (sorted []string, alreadySorted bool) {
-	groups := groupLines(b.lines, b.opts)
+	alreadySorted = true
+
+	// Sort the nested blocks first so that their changes are visible to the
+	// outer block.
+	type nestedResult struct {
+		lines         []string
+		alreadySorted bool
+	}
+	var nestedResults []nestedResult
+	for _, n := range b.nestedBlocks {
+		lines, already := n.sorted()
+		if !already {
+			alreadySorted = false
+		}
+		nestedResults = append(nestedResults, nestedResult{lines, already})
+	}
+
+	lines := b.lines
+	if !alreadySorted {
+		var lineChunks [][]string
+		// The total number of lines in lineChunks.
+		var numLines int
+		// The current position of the cursor within lines.
+		var cursor int
+		for i, nested := range b.nestedBlocks {
+			res := nestedResults[i]
+			if res.alreadySorted {
+				continue
+			}
+
+			offset := nested.start - b.start
+			// Unchanged prefix of lines.
+			lineChunks = append(lineChunks, lines[cursor:offset])
+			numLines += offset - cursor
+			// The piece of the nested block that changed.
+			lineChunks = append(lineChunks, res.lines)
+			numLines += len(res.lines)
+
+			// Advance cursor to the end of the nested block within lines.
+			cursor = offset + len(nested.lines)
+		}
+
+		if rem := lines[cursor:]; len(rem) > 0 {
+			lineChunks = append(lineChunks, rem)
+			numLines += len(rem)
+		}
+
+		lines = make([]string, 0, numLines)
+		for _, chunk := range lineChunks {
+			lines = append(lines, chunk...)
+		}
+	}
+
+	groups := groupLines(lines, b.opts)
```

Comment:
I've found one situation internally where someone is trying to do

```
// keep-sorted start block=yes

// abcd
// keep-sorted start block=yes
block_line1 {}
block_line2 {}
// keep-sorted end

// efgh
// keep-sorted start block=yes
block_line_a {}
block_line_b {}
// keep-sorted end

// keep-sorted end
```

which doesn't work all that great because the outer-most keep-sorted block doesn't do a very good job of grouping the inner keep-sorted blocks together (in fact, it doesn't even try).

 - Is this something we want to support?
 - What's the best way to handle this, if we do?
   - Maybe the grouping logic needs to change so that internal keep-sorted blocks count a group, unless they're already subsumed by a `group=yes` or `block=yes` group

Reply:
Interesting. Do you mean, like, the `abcd` and `efgh` comment is what they wanted to sort on? I can actually see demand for this, mostly in the form of gigantic config files where you might have multiple structs for each team, where the nested keep sorted (or some other start/end indicator) could be used to keep the teams sorted.

But... comments aren't part of the sorting (unless they used `sticky_comments=no`, which just breaks it in a different way). So to support this use case, we'd need a way to make comments count. Unless I'm misunderstanding the example.

Either way, it seems well out of scope as this doesn't seem to be how block sorting works.

> Maybe the grouping logic needs to change so that internal keep-sorted blocks count a group

IMO that makes sense.